core/src/main/java/com/datastax/oss/driver/api/core/paging/OffsetPager.java

```diff
@@ -0,0 +1,316 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.core.paging;
+
+import com.datastax.oss.driver.api.core.AsyncPagingIterable;
+import com.datastax.oss.driver.api.core.MappedAsyncPagingIterable;
+import com.datastax.oss.driver.api.core.PagingIterable;
+import com.datastax.oss.driver.api.core.cql.ResultSet;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import net.jcip.annotations.ThreadSafe;
+
+/**
+ * A utility to emulate offset queries on the client side (<b>this comes with important performance
+ * trade-offs, make sure you read and understand the whole javadocs before using this class</b>).
+ *
+ * <p>Web UIs and REST services often provide paginated results with random access, for example:
+ * given a page size of 20 elements, fetch page 5.
+ *
+ * <p>Cassandra does not support this natively (see <a
+ * href="https://issues.apache.org/jira/browse/CASSANDRA-6511">CASSANDRA-6511</a>), because such
+ * queries are inherently linear: the database would have to restart from the beginning every time,
+ * and skip unwanted rows until it reaches the desired offset.
+ *
+ * <p>However, from a pragmatic point of view, offset queries have their place: random pagination is
+ * a real need for many applications, and linear performance can be acceptable if the cardinality
+ * stays reasonably low.
+ *
+ * <h3>Performance considerations</h3>
+ *
+ * In concrete terms, this class <em>does</em> start from the beginning and skip rows until it
+ * reaches the desired page, every time. You need to re-execute the query in order to start with a
+ * fresh result set for each invocation:
+ *
+ * <pre>
+ * String query = "SELECT ...";
+ * OffsetPager pager = new OffsetPager(20);
+ *
+ * ResultSet rs = session.execute(query);
+ * // Iterate through rows 1-20 and discard them, then return rows 21-40
+ * OffsetPager.Page&lt;Row&gt; page2 = pager.getPage(rs, 2);
+ *
+ * rs = session.execute(query);
+ * // Iterate through rows 1-80 and discard them, then return rows 81-100
+ * OffsetPager.Page&lt;Row&gt; page5 = pager.getPage(rs, 5);
+ * </pre>
+ *
+ * This is fine for the values typically encountered in real-world applications: for example, if the
+ * page size is 25 and users never go past page 10, the worst case is only 250 rows, which is a very
```

Comment:
Actually the number of queries you have to do to get to page N, navigating one by one, is `N * (N + 1) / 2`, that is, to get at page 10 starting from page 1, you are going to execute 1 + 2 + ... + 9 + 10 = 55 queries. You probably did this math already, but was wondering if it's worth mentioning this.

Reply:
I see what you mean.
I think it's better to stay generic and describe what happens for one query, in practice clients might not always start from page 1 and click through.